File: taurus-protect-sdk-typescript/src/mappers/address.ts
/**
 * Address mapper for converting OpenAPI DTOs to domain models.
 */

import type { TgvalidatordAddress, TgvalidatordAddressAttribute } from "../internal/openapi";
import type { Address, AddressAttribute, Balance } from "../models/address";
import { safeString, safeStringDefault, safeBoolDefault, safeDate, safeMap } from "./base";

/**
 * Convert OpenAPI TgvalidatordAddressAttribute to domain AddressAttribute.
 *
 * @param dto - OpenAPI address attribute DTO
 * @returns Domain AddressAttribute model
 */
export function addressAttributeFromDto(dto: TgvalidatordAddressAttribute): AddressAttribute {
  return {
    id: safeStringDefault(dto.id, ""),
    key: safeStringDefault(dto.key, ""),
    value: safeStringDefault(dto.value, ""),
  };
}

/**
 * Convert OpenAPI TgvalidatordAddress to domain Address.
 *
 * @param dto - OpenAPI address DTO (TgvalidatordAddress)
 * @returns Domain Address model or undefined if dto is null/undefined
 */
export function addressFromDto(dto: TgvalidatordAddress | null | undefined): Address | undefined {
  if (dto == null) {
    return undefined;
  }

  // Extract attributes if present
  const attributes = safeMap(dto.attributes, addressAttributeFromDto);

  // Extract linked whitelisted address IDs
  const linkedWhitelistedAddressIds = dto.linkedWhitelistedAddressIds ?? [];

  return {
    id: safeStringDefault(dto.id, ""),
    walletId: safeStringDefault(dto.walletId, ""),
    address: safeStringDefault(dto.address, ""),
    alternateAddress: safeString(dto.alternateAddress),
    label: safeString(dto.label),
    comment: safeString(dto.comment),
    currency: safeStringDefault(dto.currency, ""),
    customerId: safeString(dto.customerId),
    externalAddressId: safeString(dto.externalAddressId),
    addressPath: safeString(dto.addressPath),
    addressIndex: safeString(dto.addressIndex),
    nonce: safeString(dto.nonce),
    status: safeString(dto.status),
    signature: safeString(dto.signature),
    disabled: safeBoolDefault(dto.disabled, false),
    canUseAllFunds: safeBoolDefault(dto.canUseAllFunds, false),
    createdAt: safeDate(dto.creationDate),
    updatedAt: safeDate(dto.updateDate),
    attributes,
    linkedWhitelistedAddressIds,
    balance: dto.balance ? {
      totalConfirmed: safeStringDefault(dto.balance.totalConfirmed, "0"),
      totalUnconfirmed: safeStringDefault(dto.balance.totalUnconfirmed, "0"),
      availableConfirmed: safeStringDefault(dto.balance.availableConfirmed, "0"),
      availableUnconfirmed: safeStringDefault(dto.balance.availableUnconfirmed, "0"),
      reservedConfirmed: safeStringDefault(dto.balance.reservedConfirmed, "0"),
      reservedUnconfirmed: safeStringDefault(dto.balance.reservedUnconfirmed, "0"),
    } : undefined,
  };
}

/**
 * Convert list of OpenAPI address DTOs to domain Addresses.
 *
 * @param dtos - List of OpenAPI address DTOs
 * @returns List of domain Address models (filters out undefined)
 */
export function addressesFromDto(
  dtos: TgvalidatordAddress[] | null | undefined
): Address[] {
  if (dtos == null) {
    return [];
  }
  const result: Address[] = [];
  for (const dto of dtos) {
    const address = addressFromDto(dto);
    if (address !== undefined) {
      result.push(address);
    }
  }
  return result;
}

File: taurus-protect-sdk-python/taurus_protect/mappers/address.py
"""Address mapper for converting OpenAPI DTOs to domain models."""

from __future__ import annotations

from typing import TYPE_CHECKING, Any, List, Optional

from taurus_protect.mappers._base import (
    safe_bool,
    safe_datetime,
    safe_list,
    safe_string,
)
from taurus_protect.mappers.wallet import balance_from_dto
from taurus_protect.models.address import Address, AddressAttribute

if TYPE_CHECKING:
    pass  # For OpenAPI types when available


def address_from_dto(dto: Any) -> Optional[Address]:
    """
    Convert OpenAPI TgvalidatordAddress to domain Address.

    Args:
        dto: OpenAPI address DTO (TgvalidatordAddress).

    Returns:
        Domain Address model or None if dto is None.
    """
    if dto is None:
        return None

    # Extract balance if present
    balance = None
    dto_balance = getattr(dto, "balance", None)
    if dto_balance is not None:
        balance = balance_from_dto(dto_balance)

    # Extract attributes if present
    attributes: List[AddressAttribute] = []
    dto_attributes = getattr(dto, "attributes", None)
    if dto_attributes is not None:
        attributes = [address_attribute_from_dto(attr) for attr in dto_attributes]

    # Extract linked whitelisted address IDs
    linked_ids = safe_list(getattr(dto, "linked_whitelisted_address_ids", None))

    return Address(
        id=safe_string(getattr(dto, "id", None)),
        wallet_id=safe_string(getattr(dto, "wallet_id", None)),
        address=safe_string(getattr(dto, "address", None)),
        alternate_address=getattr(dto, "alternate_address", None),
        label=getattr(dto, "label", None),
        comment=getattr(dto, "comment", None),
        currency=safe_string(getattr(dto, "currency", None)),
        customer_id=getattr(dto, "customer_id", None),
        external_address_id=getattr(dto, "external_address_id", None),
        address_path=getattr(dto, "address_path", None),
        address_index=getattr(dto, "address_index", None),
        nonce=getattr(dto, "nonce", None),
        status=getattr(dto, "status", None),
        balance=balance,
        signature=getattr(dto, "signature", None),
        disabled=safe_bool(getattr(dto, "disabled", None)),
        can_use_all_funds=safe_bool(getattr(dto, "can_use_all_funds", None)),
        created_at=safe_datetime(getattr(dto, "creation_date", None)),
        updated_at=safe_datetime(getattr(dto, "update_date", None)),
        attributes=attributes,
        linked_whitelisted_address_ids=linked_ids,
    )


def addresses_from_dto(dtos: Optional[List[Any]]) -> List[Address]:
    """
    Convert list of OpenAPI address DTOs to domain Addresses.

    Args:
        dtos: List of OpenAPI address DTOs.

    Returns:
        List of domain Address models.
    """
    if dtos is None:
        return []
    return [a for dto in dtos if (a := address_from_dto(dto)) is not None]


def address_attribute_from_dto(dto: Any) -> AddressAttribute:
    """
    Convert OpenAPI TgvalidatordAddressAttribute to domain AddressAttribute.

    Args:
        dto: OpenAPI address attribute DTO.

    Returns:
        Domain AddressAttribute model.
    """
    return AddressAttribute(
        id=safe_string(getattr(dto, "id", None)),
        key=safe_string(getattr(dto, "key", None)),
        value=safe_string(getattr(dto, "value", None)),
    )
